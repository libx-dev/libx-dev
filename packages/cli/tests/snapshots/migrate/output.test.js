/**
 * migrate from-libx コマンドの出力スナップショットテスト
 *
 * レジストリ出力のスナップショットテスト
 */

import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { resolve, join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { existsSync, mkdirSync, rmSync, readFileSync } from 'fs';
import migrateFromLibx from '../../../src/commands/migrate/from-libx.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const FIXTURES_DIR = resolve(__dirname, '../../fixtures/migrate-from-libx/sample-small/apps/test-project');
const TEMP_DIR = resolve(__dirname, '../../tmp/migrate-snapshots');

describe('migrate from-libx (スナップショット)', () => {
  beforeEach(() => {
    // 一時ディレクトリを作成
    if (!existsSync(TEMP_DIR)) {
      mkdirSync(TEMP_DIR, { recursive: true });
    }
  });

  afterEach(() => {
    // 一時ディレクトリをクリーンアップ
    if (existsSync(TEMP_DIR)) {
      rmSync(TEMP_DIR, { recursive: true, force: true });
    }
  });

  it('レジストリ出力が期待される構造と一致する', async () => {
    const targetPath = join(TEMP_DIR, 'docs-snapshot.json');

    const globalOpts = {
      dryRun: false,
      verbose: false,
    };

    const cmdOpts = {
      source: FIXTURES_DIR,
      projectId: 'test-project',
      target: targetPath,
      topPage: '/nonexistent/path',
      backup: join(TEMP_DIR, '.backups'),
    };

    await migrateFromLibx(globalOpts, cmdOpts);

    const registry = JSON.parse(readFileSync(targetPath, 'utf-8'));

    // 動的な値（日時、ハッシュ）を除外してスナップショット比較
    const normalizedRegistry = normalizeRegistry(registry);

    expect(normalizedRegistry).toMatchSnapshot();
  });

  it('プロジェクト構造のスナップショット', async () => {
    const targetPath = join(TEMP_DIR, 'docs-project-snapshot.json');

    const globalOpts = {
      dryRun: false,
      verbose: false,
    };

    const cmdOpts = {
      source: FIXTURES_DIR,
      projectId: 'test-project',
      target: targetPath,
      topPage: '/nonexistent/path',
      backup: join(TEMP_DIR, '.backups'),
    };

    await migrateFromLibx(globalOpts, cmdOpts);

    const registry = JSON.parse(readFileSync(targetPath, 'utf-8'));
    const project = registry.projects[0];

    // プロジェクト構造のスナップショット（コンテンツメタを除く）
    const projectStructure = {
      id: project.id,
      displayName: project.displayName,
      description: project.description,
      languages: project.languages,
      versions: project.versions,
      categories: project.categories.map(cat => ({
        id: cat.id,
        order: cat.order,
        name: cat.name,
        description: cat.description,
      })),
      documents: project.documents.map(doc => ({
        id: doc.id,
        slug: doc.slug,
        title: doc.title,
        description: doc.description,
        versions: doc.versions,
      })),
    };

    expect(projectStructure).toMatchSnapshot();
  });

  it('カテゴリとドキュメントの関連構造のスナップショット', async () => {
    const targetPath = join(TEMP_DIR, 'docs-structure-snapshot.json');

    const globalOpts = {
      dryRun: false,
      verbose: false,
    };

    const cmdOpts = {
      source: FIXTURES_DIR,
      projectId: 'test-project',
      target: targetPath,
      topPage: '/nonexistent/path',
      backup: join(TEMP_DIR, '.backups'),
    };

    await migrateFromLibx(globalOpts, cmdOpts);

    const registry = JSON.parse(readFileSync(targetPath, 'utf-8'));
    const project = registry.projects[0];

    // カテゴリとドキュメントの関連構造
    const structure = project.categories.map(cat => ({
      categoryId: cat.id,
      categoryName: cat.name,
      docs: cat.docs,
    }));

    expect(structure).toMatchSnapshot();
  });
});

/**
 * レジストリを正規化（動的な値を除外）
 *
 * @param {Object} registry - レジストリデータ
 * @returns {Object} 正規化されたレジストリデータ
 */
function normalizeRegistry(registry) {
  const normalized = JSON.parse(JSON.stringify(registry));

  // メタデータの日時を除外
  if (normalized.metadata) {
    delete normalized.metadata.createdAt;
    delete normalized.metadata.lastUpdated;
  }

  // プロジェクトごとに正規化
  if (normalized.projects) {
    for (const project of normalized.projects) {
      // ドキュメントのコンテンツメタを正規化
      if (project.documents) {
        for (const doc of project.documents) {
          if (doc.content) {
            for (const lang in doc.content) {
              const content = doc.content[lang];
              // 動的な値を除外
              delete content.syncHash;
              delete content.lastUpdated;
              delete content.reviewer;
              if (content.source) {
                delete content.source.commit;
                delete content.source.repository;
              }
            }
          }
        }
      }
    }
  }

  return normalized;
}
