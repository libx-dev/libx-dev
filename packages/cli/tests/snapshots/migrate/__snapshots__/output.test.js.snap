// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate from-libx (スナップショット) > カテゴリとドキュメントの関連構造のスナップショット 1`] = `
[
  {
    "categoryId": "guide",
    "categoryName": undefined,
    "docs": [
      "getting-started",
      "installation",
    ],
  },
]
`;

exports[`migrate from-libx (スナップショット) > プロジェクト構造のスナップショット 1`] = `
{
  "categories": [
    {
      "description": {
        "en": "",
        "ja": "",
      },
      "id": "guide",
      "name": undefined,
      "order": 1,
    },
  ],
  "description": {
    "en": "A test project for migration testing",
    "ja": "マイグレーションテスト用のプロジェクト",
  },
  "displayName": {
    "en": "Test Project",
    "ja": "テストプロジェクト",
  },
  "documents": [
    {
      "description": undefined,
      "id": "getting-started",
      "slug": "guide/getting-started",
      "title": {
        "en": "Getting Started",
        "ja": "はじめに",
      },
      "versions": [
        "v1",
      ],
    },
    {
      "description": undefined,
      "id": "installation",
      "slug": "guide/installation",
      "title": {
        "en": "Installation",
        "ja": "インストール",
      },
      "versions": [
        "v1",
      ],
    },
  ],
  "id": "test-project",
  "languages": [
    {
      "code": "en",
      "default": true,
      "displayName": "English",
      "status": "active",
    },
    {
      "code": "ja",
      "default": false,
      "displayName": "日本語",
      "fallback": "en",
      "status": "active",
    },
  ],
  "versions": [
    {
      "date": "2024-01-01T00:00:00.000Z",
      "id": "v1",
      "isLatest": true,
      "name": "Version 1.0",
      "status": "active",
    },
  ],
}
`;

exports[`migrate from-libx (スナップショット) > レジストリ出力が期待される構造と一致する 1`] = `
{
  "$schemaVersion": "1.0.0",
  "metadata": {
    "createdBy": "migrate-from-libx",
  },
  "projects": [
    {
      "categories": [
        {
          "description": {
            "en": "",
            "ja": "",
          },
          "docs": [
            "getting-started",
            "installation",
          ],
          "id": "guide",
          "order": 1,
          "titles": {
            "en": "Guide",
            "ja": "ガイド",
          },
        },
      ],
      "description": {
        "en": "A test project for migration testing",
        "ja": "マイグレーションテスト用のプロジェクト",
      },
      "displayName": {
        "en": "Test Project",
        "ja": "テストプロジェクト",
      },
      "documents": [
        {
          "content": {
            "en": {
              "path": "content/getting-started/en.mdx",
              "source": {
                "branch": "main",
                "path": "src/content/docs/v1/en/01-guide/01-getting-started.mdx",
              },
              "status": "published",
              "wordCount": 34,
            },
            "ja": {
              "path": "content/getting-started/ja.mdx",
              "source": {
                "branch": "main",
                "path": "src/content/docs/v1/ja/01-guide/01-getting-started.mdx",
              },
              "status": "published",
              "wordCount": 13,
            },
          },
          "id": "getting-started",
          "keywords": [],
          "slug": "guide/getting-started",
          "status": "published",
          "summary": {
            "en": "Quick start guide for Test Project",
            "ja": "Test Projectのクイックスタートガイド",
          },
          "tags": [],
          "title": {
            "en": "Getting Started",
            "ja": "はじめに",
          },
          "versions": [
            "v1",
          ],
          "visibility": "public",
        },
        {
          "content": {
            "en": {
              "path": "content/installation/en.mdx",
              "source": {
                "branch": "main",
                "path": "src/content/docs/v1/en/01-guide/02-installation.mdx",
              },
              "status": "published",
              "wordCount": 27,
            },
            "ja": {
              "path": "content/installation/ja.mdx",
              "source": {
                "branch": "main",
                "path": "src/content/docs/v1/ja/01-guide/02-installation.mdx",
              },
              "status": "published",
              "wordCount": 12,
            },
          },
          "id": "installation",
          "keywords": [],
          "slug": "guide/installation",
          "status": "published",
          "summary": {
            "en": "How to install Test Project",
            "ja": "Test Projectのインストール方法",
          },
          "tags": [],
          "title": {
            "en": "Installation",
            "ja": "インストール",
          },
          "versions": [
            "v1",
          ],
          "visibility": "public",
        },
      ],
      "id": "test-project",
      "languages": [
        {
          "code": "en",
          "default": true,
          "displayName": "English",
          "status": "active",
        },
        {
          "code": "ja",
          "default": false,
          "displayName": "日本語",
          "fallback": "en",
          "status": "active",
        },
      ],
      "licenses": [
        {
          "author": "Test Team",
          "id": "test-original",
          "license": "MIT",
          "name": "Test Documentation",
          "url": "https://opensource.org/licenses/MIT",
        },
      ],
      "versions": [
        {
          "date": "2024-01-01T00:00:00.000Z",
          "id": "v1",
          "isLatest": true,
          "name": "Version 1.0",
          "status": "active",
        },
      ],
    },
  ],
  "settings": {
    "defaultLocale": "en",
    "siteName": "Documentation",
    "siteUrl": "https://example.com",
  },
}
`;
